catalog/models_operations.py

```diff
@@ -1,57 +1,134 @@
+
+
+class AddIndex(IndexSQL):
+
+    def __init__(self, name: str, columns: typing.List[str]):
+        self.name = name
+        self.columns = columns
+
+    def execute(self, table, schema_editor):
+        schema_editor.execute(
+            f'CREATE INDEX {self._index_name(table)} ON {table}'
+            f'({", ".join(self.columns)});'
+        )
+
+
+class DropIndex(IndexSQL):
+
+    def execute(self, table, schema_editor):
+        schema_editor.execute(
+            f'DROP INDEX {self._index_name(table)};'
+        )
+
+
+class IndexOperation(Operation):
     """
-    Create a B-Tree index by given columns or expressions.
+    Operate an index by given IndexSQL objects.
 
     Docs: https://docs.djangoproject.com/en/1.11/ref/migration-operations/#writing-your-own
     """
 
     reduces_to_sql = True
     reversible = True
 
-    def __init__(self, name, columns, model_name):
-        self.name = name
-        self.columns = columns
+    def __init__(self, model_name, forward: IndexSQL, backward: IndexSQL):
         self.model_name = model_name
+        self.forward = forward
+        self.backward = backward
 
     def state_forwards(self, app_label, state):
         """We have to implement this method for Operation interface."""
 
-    def _index_name(self, table_name):
-        return f'{table_name}_{self.name}_idx'
-
     def database_forwards(self, app_label, schema_editor, _, to_state):
         to_model = to_state.apps.get_model(app_label, self.model_name)
         if self.allow_migrate_model(schema_editor.connection.alias, to_model):
             table_name = to_model._meta.db_table
-            schema_editor.execute(
-                f'CREATE INDEX {self._index_name(table_name)} ON {table_name}'
-                f'({", ".join(self.columns)});'
-            )
+            self.forward.execute(table_name, schema_editor)
 
     def database_backwards(self, app_label, schema_editor, from_state, _):
         from_model = from_state.apps.get_model(app_label, self.model_name)
         if self.allow_migrate_model(schema_editor.connection.alias, from_model):
             table_name = from_model._meta.db_table
-            schema_editor.execute(
-                f'DROP INDEX {self._index_name(table_name)};'
-            )
+            self.backward.execute(table_name, schema_editor)
 
     def describe(self):
-        return f'Create index {self.name} for {self.model_name}'
+        return f'Operate the index {self.name} for {self.model_name}'
+
+
+class RevertOperation(Operation):
+
+    reduces_to_sql = True
+    reversible = True
+
+    def __init__(self, operation: IndexOperation):
+        self.operation = operation
+
+    def state_forwards(self, app_label, state):
+        """We have to implement this method for Operation interface."""
+
+    def database_forwards(self, app_label, schema_editor, from_state, to_state):
+        self.operation.database_backwards(app_label, schema_editor, from_state, to_state)
```

Comment:
```suggestion
        self.operation.database_forwards(app_label, schema_editor, from_state, to_state)
```

Reply:
The implementation is correct. It allows to revert wrapped operation